syz-manager/covfilter.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"regexp"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/mgrconfig"
+	"github.com/google/syzkaller/sys/targets"
+)
+
+type CoverFilter struct {
+	pcStart     uint32
+	pcSize      uint32
+	pcEnd       uint32
+	weightedPCs map[uint32]float32
+
+	bitmapFilename string
+	target         *targets.Target
+}
+
+func createCoverageFilter(cfg *mgrconfig.Config, target *targets.Target) (covFilterFilename string, err error) {
+	files := cfg.CovFilter.Files
+	funcs := cfg.CovFilter.Functions
+	rawPCs := cfg.CovFilter.RawPCs
+	if len(files) == 0 && len(funcs) == 0 && len(rawPCs) == 0 {
+		return "", nil
+	}
+	filesRegexp, err := getRegexps(files)
+	if err != nil {
+		return "", err
+	}
+	funcsRegexp, err := getRegexps(funcs)
+	if err != nil {
+		return "", err
+	}
+	if len(filesRegexp) > 0 || len(funcsRegexp) > 0 {
+		log.Logf(0, "initialize coverage information...")
+		if err = initCover(target, cfg.KernelObj, cfg.KernelSrc, cfg.KernelBuildSrc); err != nil {
+			return "", err
+		}
+	}
+
+	covFilter := CoverFilter{
+		weightedPCs:    make(map[uint32]float32),
+		target:         target,
+		bitmapFilename: cfg.Workdir + "/" + "syz-cover-bitmap",
+	}
+	err = covFilter.initWeightedPCs(filesRegexp, funcsRegexp, rawPCs)
+	if err != nil {
+		return "", err
+	}
+	err = covFilter.createBitmap()
+	if err != nil {
+		return "", err
+	}
+	return covFilter.bitmapFilename, nil
+}
+
+func (covFilter *CoverFilter) initWeightedPCs(filesRegexp, funcsRegexp []regexp.Regexp, rawPCsFiles []string) error {
+	for _, f := range rawPCsFiles {
+		rawFile, err := os.Open(f)
```

Comment:
Ideally, *os.File is wrapper into bufio.Buffer if it's used for fine-grained reads. Otherwise each byte read that fmt.Fscanf below will issue will translate into read system call. See:
https://play.golang.org/p/w_jSd3pe2XJ